From c5475545e93f035a88fd743e3f7dfb69affee6b4 Mon Sep 17 00:00:00 2001
From: Aaron Kling <webgeek1234@gmail.com>
Date: Sat, 14 May 2022 13:28:39 -0500
Subject: [PATCH 4/5] Give mediaprovider_app mlstrustedsubject

This is necessary to allow writing to an external sdcard with ext4

Give mediaprovider_app access to priv_app data so it can access its
own app data

Change-Id: I1d7ce713cd5f3d8c3bb4251c5ec06793e538be43
---
 prebuilts/api/33.0/private/app_neverallows.te   | 2 +-
 prebuilts/api/33.0/private/mediaprovider_app.te | 5 ++++-
 prebuilts/api/33.0/private/mlstrustedsubject.te | 3 +++
 private/app_neverallows.te                      | 2 +-
 private/mediaprovider_app.te                    | 5 ++++-
 private/mlstrustedsubject.te                    | 3 +++
 6 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/prebuilts/api/33.0/private/app_neverallows.te b/prebuilts/api/33.0/private/app_neverallows.te
index 304f5a209..9a081cf57 100644
--- a/prebuilts/api/33.0/private/app_neverallows.te
+++ b/prebuilts/api/33.0/private/app_neverallows.te
@@ -83,7 +83,7 @@ neverallow {
 # permission only makes sense within a domain (hence should
 # never be granted to any other domain within mlstrustedsubject)
 # and an untrusted app is allowed fork permission to itself.
-neverallow all_untrusted_apps mlstrustedsubject:process fork;
+neverallow { all_untrusted_apps -mediaprovider_app } mlstrustedsubject:process fork;
 
 # Do not allow untrusted apps to hard link to any files.
 # In particular, if an untrusted app links to other app data
diff --git a/prebuilts/api/33.0/private/mediaprovider_app.te b/prebuilts/api/33.0/private/mediaprovider_app.te
index a9a52bbe8..b9a7efa4c 100644
--- a/prebuilts/api/33.0/private/mediaprovider_app.te
+++ b/prebuilts/api/33.0/private/mediaprovider_app.te
@@ -1,7 +1,7 @@
 ###
 ### A domain for further sandboxing the MediaProvider mainline module.
 ###
-type mediaprovider_app, domain, coredomain, bpfdomain;
+type mediaprovider_app, domain, coredomain, bpfdomain, mlstrustedsubject;
 
 app_domain(mediaprovider_app)
 
@@ -68,3 +68,6 @@ dontaudit mediaprovider_app sysfs_vendor_sched:file w_file_perms;
 # bpfprog access for FUSE BPF
 allow mediaprovider_app fs_bpf:file read;
 allow mediaprovider_app bpfloader:bpf { map_read map_write prog_run };
+
+allow mediaprovider_app privapp_data_file:dir create_dir_perms;
+allow mediaprovider_app privapp_data_file:file create_file_perms;
diff --git a/prebuilts/api/33.0/private/mlstrustedsubject.te b/prebuilts/api/33.0/private/mlstrustedsubject.te
index 22482d9b7..32079bbe3 100644
--- a/prebuilts/api/33.0/private/mlstrustedsubject.te
+++ b/prebuilts/api/33.0/private/mlstrustedsubject.te
@@ -9,6 +9,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
 } { app_data_file privapp_data_file }:file ~{ read write map getattr ioctl lock append };
 
 neverallow {
@@ -16,6 +17,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
 } { app_data_file privapp_data_file }:dir ~{ read getattr search };
 
 neverallow {
@@ -23,6 +25,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
   -system_server
   -adbd
   -runas
diff --git a/private/app_neverallows.te b/private/app_neverallows.te
index 304f5a209..9a081cf57 100644
--- a/private/app_neverallows.te
+++ b/private/app_neverallows.te
@@ -83,7 +83,7 @@ neverallow {
 # permission only makes sense within a domain (hence should
 # never be granted to any other domain within mlstrustedsubject)
 # and an untrusted app is allowed fork permission to itself.
-neverallow all_untrusted_apps mlstrustedsubject:process fork;
+neverallow { all_untrusted_apps -mediaprovider_app } mlstrustedsubject:process fork;
 
 # Do not allow untrusted apps to hard link to any files.
 # In particular, if an untrusted app links to other app data
diff --git a/private/mediaprovider_app.te b/private/mediaprovider_app.te
index a9a52bbe8..b9a7efa4c 100644
--- a/private/mediaprovider_app.te
+++ b/private/mediaprovider_app.te
@@ -1,7 +1,7 @@
 ###
 ### A domain for further sandboxing the MediaProvider mainline module.
 ###
-type mediaprovider_app, domain, coredomain, bpfdomain;
+type mediaprovider_app, domain, coredomain, bpfdomain, mlstrustedsubject;
 
 app_domain(mediaprovider_app)
 
@@ -68,3 +68,6 @@ dontaudit mediaprovider_app sysfs_vendor_sched:file w_file_perms;
 # bpfprog access for FUSE BPF
 allow mediaprovider_app fs_bpf:file read;
 allow mediaprovider_app bpfloader:bpf { map_read map_write prog_run };
+
+allow mediaprovider_app privapp_data_file:dir create_dir_perms;
+allow mediaprovider_app privapp_data_file:file create_file_perms;
diff --git a/private/mlstrustedsubject.te b/private/mlstrustedsubject.te
index 22482d9b7..32079bbe3 100644
--- a/private/mlstrustedsubject.te
+++ b/private/mlstrustedsubject.te
@@ -9,6 +9,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
 } { app_data_file privapp_data_file }:file ~{ read write map getattr ioctl lock append };
 
 neverallow {
@@ -16,6 +17,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
 } { app_data_file privapp_data_file }:dir ~{ read getattr search };
 
 neverallow {
@@ -23,6 +25,7 @@ neverallow {
   -installd
   -iorap_prefetcherd
   -iorap_inode2filename
+  -mediaprovider_app
   -system_server
   -adbd
   -runas
-- 
2.30.2

